Histogram viz yaxis range

moontrip opened the pull request:
This PR resolves disappearance of y-axis tick labels for certain variables at histogram viz. I have made screenshots for before and after setting y-axis range: note that I captured screenshots before merging main branch so white gap between bar is still showing 😃 

### - before this work: both count and proportion with log scale (GEMS1 Case Control data)
![histogramViz-noRange-count-log](https://user-images.githubusercontent.com/12802305/137202931-e3377d52-171e-4fb9-8506-7e6f0963149c.png)

![histogramViz-noRange-proportion-log](https://user-images.githubusercontent.com/12802305/137202945-fbd4a8ef-1c05-4569-8154-39a9524df086.png)

### - after applying this work: both count and proportion with log scale (GEMS1 Case Control data
![histogramViz-Range-count-log](https://user-images.githubusercontent.com/12802305/137202992-b2783dbe-79ab-41e5-b84b-48af2887b04a.png)

![histogramViz-Range-proportion-log](https://user-images.githubusercontent.com/12802305/137203003-50457325-e6e5-46ea-b2fb-af19602fc4be.png)

bobular commented on src/lib/core/components/visualizations/implementations/HistogramVisualization.tsx:
```diff
@@ -275,6 +278,24 @@ function HistogramViz(props: VisualizationProps) {
+      data.value && data.value.series.length > 0
+        ? findMaxOfStackedArray(data.value.series)
+        : undefined,
+    [data, xAxisVariable, overlayVariable]
```
Do we need `xAxisVariable, overlayVariable` in the dependencies?

jtlong3rd commented on src/lib/core/components/visualizations/implementations/HistogramVisualization.tsx:
```diff
@@ -639,3 +662,62 @@ function reorderData(
+      let tempArray = [];
+      // looping through each data array
+      for (let j = 0; j < uniqueBinStart.length; j++) {
+        // check if binStart exists compared to unique binStart array
+        if (dataArray[i].binStart.includes(uniqueBinStart[j])) {
+          let indexCount = dataArray[i].binStart.indexOf(uniqueBinStart[j]);
+          // make tempArray for data.count
+          tempArray.push(dataArray[i].count[indexCount]);
+        } else {
+          // add filler count, 0 for non-existence case
+          tempArray.push(0);
+        }
```
Slight simplification using `findIndex`:

```suggestion
        // check if binStart exists compared to unique binStart array
        const indexCount = dataArray[i].binStart.findIndex(bStart => bStart ===  uniqueBinStart[j]);
        
        if (indexCount >= 0) {
          // make tempArray for data.count
          tempArray.push(dataArray[i].count[indexCount]);
        } else {
          // add filler count, 0 for non-existence case
          tempArray.push(0);
        }
```

jtlong3rd commented on src/lib/core/components/visualizations/implementations/HistogramVisualization.tsx:
```diff
@@ -639,3 +662,62 @@ function reorderData(
+  // construct 2D totalCountArray
+  // [# of stratification] rows and [# of unique binStart] at each low
+  if (uniqueBinStart != null) {
+    for (let i = 0; i < dataArray.length; i++) {
```
A [for...of](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...ofl) loop might improve readability.

jtlong3rd commented on src/lib/core/components/visualizations/implementations/HistogramVisualization.tsx:
```diff
@@ -639,3 +662,62 @@ function reorderData(
+      // temporary array to store each row
+      let tempArray = [];
+      // looping through each data array
+      for (let j = 0; j < uniqueBinStart.length; j++) {
```
A [for...of](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...ofl) loop might improve readability.

moontrip commented:
@bobular Thank you for your reivew! 👍 I have addressed two issues you found. Removed unnecessary dependencies (for both histogram and barplot viz) and set a condition to check vizConfig.dependentAxisLogScale to set range min of 0 (non-log) or 0.001 (log) for proportion

jtlong3rd commented:
> @jtlong3rd Thank you for your review and suggestions! 👍 May I ask one question regarding `for-of` for clarification ? so for the example with the first for loop, i.e., `for (let i = 0; i < dataArray.length; i++)`
> 
> Did you have in mind using like this ? `for (const i of dataArray.keys())`

Almost! You can actually iterate over the elements of the dataArray like this: `for (const dataArrayElement of dataArray) {`. You can even do this: `for (const { binStart, count } of dataArray) {`.

bobular commented:
@moontrip  I'm not usually a fan of reduce (right, Jamie?!), but I think it's what's needed for the max-finder function

This returns the exact same result as your function, on one plot I tested with.
```
function findMaxOfStackedArray(
  data: HistogramDataSeries[]
) {
  // calculate the sum of all the counts from bins with the same label
  const sumsByLabel = data
    .flatMap( // make an array of [ [ label, count ], [ label, count ], ... ] from all series
      (series) => zip(series.bins.map((bin) => bin.binLabel),
		      series.bins.map((bin) => bin.count)))
      // then do a sum of counts per label
    .reduce<Record<string, number>>(
      (map, [ label, count ]) => {
	if (label != null && count != null) { // zip can potentially return nulls
	  if (map[label] == null)
	    map[label] = 0;
	  map[label] = map[label] + count;
	}
	return map;
      },
      // empty map for reduce to start with
      {}
    );

  return max(Object.values(sumsByLabel));
}
```

bobular commented on src/lib/core/components/visualizations/implementations/HistogramVisualization.tsx:
```diff
@@ -11,12 +11,15 @@ import {
 import { pipe } from 'fp-ts/lib/function';
 import * as t from 'io-ts';
-import { isEqual } from 'lodash';
+import { isEqual, sortBy, uniq, flatMap, max } from 'lodash';
```
You'd need `zip` here, but no need for `flatMap` it's standard JS the way we use it.

jtlong3rd commented:
> @moontrip I'm not usually a fan of reduce (right, Jamie?!), but I think it's what's needed for the max-finder function
> 
> This returns the exact same result as your function, on one plot I tested with.
> 
> ```
> function findMaxOfStackedArray(
>   data: HistogramDataSeries[]
> ) {
>   // calculate the sum of all the counts from bins with the same label
>   const sumsByLabel = data
>     .flatMap( // make an array of [ [ label, count ], [ label, count ], ... ] from all series
>       (series) => zip(series.bins.map((bin) => bin.binLabel),
> 		      series.bins.map((bin) => bin.count)))
>       // then do a sum of counts per label
>     .reduce<Record<string, number>>(
>       (map, [ label, count ]) => {
> 	if (label != null && count != null) { // zip can potentially return nulls
> 	  if (map[label] == null)
> 	    map[label] = 0;
> 	  map[label] = map[label] + count;
> 	}
> 	return map;
>       },
>       // empty map for reduce to start with
>       {}
>     );
> 
>   return max(Object.values(sumsByLabel));
> }
> ```

Slick! How about `(series) => series.bins.map(bin => [bin.binLabel, bin.count])` instead of `(series) => zip(series.bins.map((bin) => bin.binLabel), series.bins.map((bin) => bin.count)))`?

jtlong3rd reviewed (approved):
🎉

bobular commented:
Cool - I think we're over the line.  :tada:  :beers:  :fireworks:
